internal/command/postgres/failover.go

```diff
@@ -128,3 +141,230 @@ func runFailover(ctx context.Context) (err error) {
+	if err != nil {
+		return err
+	}
+	if mach.Status == "error" {
+		return fmt.Errorf("old leader %s could not be started: %s", leader.ID, mach.Message)
+	}
+
+	fmt.Printf("Waiting for leadership to swap to %s...\n", newLeader.ID)
+	if err := retry.Do(
+		func() error {
+			leader, err = flapsClient.Get(ctx, newLeader.ID)
+			if err != nil {
+				return err
+			}
+
+			if isLeader(leader) {
+				return nil
+			} else {
+				return fmt.Errorf("Machine %s never became the leader", newLeader.ID)
+			}
+
+		},
+		retry.Context(ctx), retry.Attempts(60), retry.Delay(time.Second), retry.DelayType(retry.FixedDelay),
+	); err != nil {
+		return err
+	}
+
+	// wait for health checks to pass
+	if err := watch.MachinesChecks(ctx, machines); err != nil {
+		return fmt.Errorf("failed to wait for health checks to pass: %w", err)
+	}
+
+	fmt.Fprintf(io.Out, "Failover complete\n")
+	return nil
+}
+
+func handleFlexFailoverFail(ctx context.Context, machines []*api.Machine) (err error) {
+	io := iostreams.FromContext(ctx)
+	flapsClient := flaps.FromContext(ctx)
+
+	leader, err := pickLeader(ctx, machines)
+	if err != nil {
+		return err
+	}
+
+	fmt.Fprintln(io.ErrOut, "Error promoting new leader, restarting existing leader")
+	fmt.Println("Waiting for old leader to finish stopping")
+	if err := retry.Do(
+		func() error {
+			leader, err = flapsClient.Get(ctx, leader.ID)
+			if err != nil {
+				return err
+			}
+
+			if leader.State == "stopped" || leader.State == "started" {
+				return nil
+			} else if leader.State == "stopping" {
+				return fmt.Errorf("Old leader hasn't finished stopping")
+			} else {
+				return fmt.Errorf("Old leader is in an unexpected state: %s", leader.State)
+			}
+
+		},
+		retry.Context(ctx), retry.Attempts(60), retry.Delay(time.Second), retry.DelayType(retry.FixedDelay),
+	); err != nil {
+		return err
+	}
+
+	fmt.Println("Clearing existing machine lease...")
+
+	// Clear the existing lease on this machine
+	lease, err := flapsClient.FindLease(ctx, leader.ID)
+	if err != nil {
+		if !strings.Contains(err.Error(), " lease not found") {
+			return err
+		}
+	}
+	if err := flapsClient.ReleaseLease(ctx, leader.ID, lease.Data.Nonce); err != nil {
+		return err
+	}
+
+	fmt.Println("Trying to start old leader")
+	// Start the machine again
+	leader, err = flapsClient.Get(ctx, leader.ID)
+	if err != nil {
+		return err
+	}
+
+	mach, err := flapsClient.Start(ctx, leader.ID, leader.LeaseNonce)
+	if err != nil {
+		return err
+	}
+	if mach.Status == "error" {
+		return fmt.Errorf("old leader %s could not be started: %s", leader.ID, mach.Message)
+	}
+
+	fmt.Println("Old leader started succesfully")
+
+	return nil
+}
+
+func pickNewLeader(ctx context.Context, app *api.AppCompact, machines_within_primary_region []*api.Machine) (*api.Machine, error) {
+	machine_reasons := make(map[string]string)
+
+	for _, machine := range machines_within_primary_region {
+		is_valid := true
+		if isLeader(machine) {
```

Comment:
The list of machines should exclude the existing leader.

Reply:
i'm a little confused, doesn't the code already exclude the existing leader?